DuckDuckGo/SpecialErrorPage/SpecialErrorPageInterfaces.swift

```diff
@@ -0,0 +1,131 @@
+    ///   - challenge: The authentication challenge.
+    ///   - completionHandler: A completion handler block to execute with the response.
+    func handleWebView(_ webView: WKWebView, didReceive challenge: URLAuthenticationChallenge, completionHandler: @escaping (URLSession.AuthChallengeDisposition, URLCredential?) -> Void)
+
+    /// Handles failures during provisional navigation.
+    ///
+    /// - Parameters:
+    ///   - webView: The `WKWebView` instance that failed the navigation.
+    ///   - navigation: The navigation object for the operation.
+    ///   - error: The error that occurred.
+    func handleWebView(_ webView: WKWebView, didFailProvisionalNavigation navigation: WebViewNavigation, withError error: NSError)
+
+    /// Handles the successful completion of a navigation in the web view.
+    ///
+    /// - Parameters:
+    ///   - webView: The web view that loaded the content.
+    ///   - navigation: The navigation object that finished.
+    func handleWebView(_ webView: WKWebView, didFinish navigation: WebViewNavigation)
+}
+
+// MARK: - SpecialErrorPageActionHandler
+
+/// A type that defines actions for handling special error pages.
+///
+/// This protocol is intended to be adopted by types that need to manage user interactions
+/// with special error pages, such as navigating to a site, leaving a site, or presenting
+/// advanced information related to the error.
+protocol SpecialErrorPageActionHandler {
+    /// Handles the action of navigating to the site associated with the error page
+    func visitSite()
+
+    /// Handles the action of leaving the site associated with the error page
+    func leaveSite()
+
+    /// Handles the action of requesting more detailed information about the error
+    func advancedInfoPresented()
+}
+
+// MARK: - BaseSpecialErrorPageNavigationHandling
+
+/// A type that defines the base functionality for handling navigation related to special error pages.
+protocol BaseSpecialErrorPageNavigationHandling: AnyObject {
+    /// The delegate that handles navigation actions for special error pages.
+    var delegate: SpecialErrorPageNavigationDelegate? { get set }
+
+    /// A Boolean value indicating whether the special error page is currently visible.
+    var isSpecialErrorPageVisible: Bool { get }
+
+    /// The URL that failed to load, if any.
+    var failedURL: URL? { get }
+
+    /// Attaches a web view to the special error page handling.
+    func attachWebView(_ webView: WKWebView)
+
+    /// Sets the user script for the special error page.
+    func setUserScript(_ userScript: SpecialErrorPageUserScript?)
+}
+
+typealias SpecialErrorPageNavigationHandling = BaseSpecialErrorPageNavigationHandling & WebViewNavigationHandling & SpecialErrorPageUserScriptDelegate
+
+// MARK: - SpecialErrorPageNavigationDelegate
+
+/// A delegate for handling navigation actions related to special error pages.
+protocol SpecialErrorPageNavigationDelegate: AnyObject {
```

Comment:
can we move this protocol above BaseSpecialError protocol where it is already used?